Support multi-targeting for Roslyn components

eerhardt opened the pull request:
Allows for Roslyn components (analyzers, source generators) to target mulltiple Roslyn API versions in a single package. The highest compatible asset is selected.

Fix #20355

eerhardt committed:
Support multi-targeting for Roslyn components

Allows for Roslyn components (analyzers, source generators) to target mulltiple Roslyn API versions in a single package. The highest compatible asset is selected.

Fix #20355

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageDependencies.cs:
```diff
@@ -114,6 +114,14 @@ public string ProjectLanguage
+        /// <summary>
+        /// Optional version of the compiler API (E.g. 'roslyn3.9', 'roslyn4.0')
+        /// </summary>
+        public string CompilerApiVersion
```
Most of the `ResolvePackageDependencies` task is legacy behavior which is disabled unless you set `EmitLegacyAssetsFileItems` to true.  See https://docs.microsoft.com/dotnet/core/compatibility/msbuild#design-time-builds-only-return-top-level-package-references.

So I don't think we need to update this task.  No one should be consuming the analyzer information that it generates, and if they do it should be OK that they don't handle multi-targeted analyzers.

eerhardt replied:
Ah, that makes sense. Thanks for the information on this. I can revert this change.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.PackageDependencyResolution.targets:
```diff
@@ -172,12 +176,28 @@ Copyright (c) .NET Foundation. All rights reserved.
             AssemblyFile="$(MicrosoftNETBuildTasksAssembly)" />
 
+  <!-- Reads the version of the compiler APIs that are currently being used in order to pick the correct Roslyn components. -->
+  <Target Name="_ResolveCompilerVersion"
```
Can we change this so that the code to read the compiler API version is done as part of the `ResolvePackageAssets` task instead of computed beforehand and passed in?  That way if we don't have any references to multi-targeted analyzers we don't even have to read the API version from the assembly, so we don't have to worry about the perf impact.

In that case probably you would need to pass in the path to the CodeAnalysis DLL to the task instead of the version, which would make testing harder.

eerhardt replied:
The idea we had offline was that the Roslyn targets would be setting the `$(CompilerApiVersion)` property statically. Thus this target wouldn't need to run at all.

cc @chsienki 

> That way if we don't have any references to multi-targeted analyzers

We are using this property to figure out if we have any references to multi-targeted analyzers at all. Using the first part of this property's value (`roslyn`), we are looking for folders that start with that name in the `analyzers` folder. I didn't want to hard-code "roslyn" into the C# task. It seems like a better design to put that information up in the .targets file, where it can be changed easily, if needed.

dsplaisted reviewed (commented):
I haven't reviewed the code in detail yet, but I did have some quick feedback.

eerhardt committed:
Respond to PR feedback.

Undo changes to RunResolvePackageDependencies, since it only affects "legacy" behavior, which shouldn't be necessary to change.

eerhardt committed:
Fix tests

eerhardt commented:
@dsplaisted @marcpopMSFT - any thoughts here? I'd like to get this into 6.0 to unblock the multi-targeting work in https://github.com/dotnet/runtime/pull/58446

ericstj commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.PackageDependencyResolution.targets:
```diff
@@ -207,9 +211,24 @@ Copyright (c) .NET Foundation. All rights reserved.
 
+  <!-- Reads the version of the compiler APIs that are currently being used in order to pick the correct Roslyn components. -->
+  <Target Name="_ResolveCompilerVersion"
+          Condition="'$(CompilerApiVersion)' == ''">
```
Do we have to think about any non-roslyn project types?  Is there something that tells us that a project type supports analyzers?

eerhardt replied:
> Do we have to think about any non-roslyn project types?

Yes. We need to consider F# projects for sure. We may also need to think about C++/CLI projects.

> Is there something that tells us that a project type supports analyzers?

I haven't found anything. I did find the `RoslynComponent` capability, but that means this project is itself an analyzer.

https://github.com/dotnet/roslyn/blob/86ff365d5d28fddd4a4738cf1d9d101522dffd50/src/Compilers/Core/MSBuildTask/Microsoft.Managed.Core.targets#L348

I can condition this on `Language == 'C#' or 'VB'` if you think that would be safer..? The Task is written so it can accept an empty/null `CompilerApiVersion`, so not setting that property for non-C#/VB projects will be safe.

Thoughts @chsienki ?

chsienki replied:
`RoslynComponent` is only used for the debugging experience, so can't rely on that. I think gating on Language is probably sensible. I'm not sure what will happen if you try and read from a non-existent assembly? If it errors you could also condition on `Condition="Exists('<tasksassembly.dll>')"` to make sure we only try and read it when it's there.

eerhardt replied:
That's a good call. I'll add both conditions.

eerhardt committed:
Respond to PR feedback

- Add more conditions to when resolving the Roslyn version runs - only for C# and VB projects, and only if the CodeAnalysis.dll file exists

chsienki commented on src/Tasks/Microsoft.NET.Build.Tasks/NuGetUtils.NuGet.cs:
```diff
@@ -47,7 +47,167 @@ public static NuGetFramework ParseFrameworkName(string frameworkName)
+        /// in the folder with the highest applicable compiler version are picked.
+        /// In this case,
+        /// 
+        /// "analyzers/dotnet/roslyn3.8/analyzer.dll"
```
Is this comment correct? For 3.9 if should pick 4.0 right?

eerhardt replied:
I believe the comment is correct.  If we are running on Roslyn v3.9, we can't pick the `roslyn4.0` asset, or else we would get a load exception that `Microsoft.CodeAnalysis.dll version=4.0.0.0 is not found`.

Since `v4.0` is higher than the current version `v3.9`, we have to exclude it.

This works very similar to TFM selection: you can't pick a `net6.0` asset out of a NuGet package when the app is trying to run on `net5.0`.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/NuGetUtils.NuGet.cs:
```diff
@@ -47,7 +47,167 @@ public static NuGetFramework ParseFrameworkName(string frameworkName)
+        /// in the folder with the highest applicable compiler version are picked.
+        /// In this case,
+        /// 
+        /// "analyzers/dotnet/roslyn3.8/analyzer.dll"
```
I don't think so, if the Roslyn compiler version is 3.9, then it can't necessarily run analyzers written for Roslyn 4.0, so it should use the ones written for Roslyn 3.8.

chsienki replied:
Yeah, ignore me I was just having a slow moment.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -848,6 +855,8 @@ public int GetHashCode((string, NuGetVersion) library)
 
             private void WriteAnalyzers()
             {
+                HashSet<string>? excludedAnalyzers = NuGetUtils.GetExcludedAnalyzers(_lockFile, _task.ProjectLanguage, _task.CompilerApiVersion);
```
I would like to see the logic to check if an analyzer applies refactored.  However, feel free to merge as is since you're crunched for time.

The way this is structured seems confusing.  `GetExcludedAnalyzers` ends up calling `IsApplicableAnalyzer`, but then returns a HashSet which is passed back in to `IsApplicableAnalyzer`.

I think it would be easier to understand (and possibly faster) if `IsApplicableAnalyzer` simply contained all the logic to determine whether an analyzer was applicable, including the multi-targeting logic, instead of separately computing a list of excluded analyzers.  If needed, that could be wrapped in a class which could maintain state to perform expensive operations once, lazily, as necessary.

eerhardt replied:
Your idea of wrapping the logic in a class allowed me to find a way to make this work.

It isn't possible to make this decision directly in `IsApplicableAnalyzer` because we need to see all the assets before we know which assets apply and which don't. Creating a class to keep track of the necessary information allowed me to track the potential assets, and then add them at the end.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.PackageDependencyResolution.targets:
```diff
@@ -207,9 +211,26 @@ Copyright (c) .NET Foundation. All rights reserved.
   </Target>
 
+  <!-- Reads the version of the compiler APIs that are currently being used in order to pick the correct Roslyn components. -->
+  <Target Name="_ResolveCompilerVersion"
```
I'd like to see the Roslyn targets updated to set this version number soon so that this target is normally never actually run.

eerhardt replied:
@chsienki - any chance you can take care of this? I can open an issue in dotnet/roslyn if you'd like.

chsienki replied:
We have an issue here https://github.com/dotnet/roslyn/issues/52265 

It's easy enough to ship in the .NET 6 timeframe by just hard coding it; my concern has always been around the maintainability and ensuring it never gets out of sync, but we can figure that out after the fact if we really need it here.

eerhardt committed:
Respond to PR feedback

Refactor analyzer resolution logic to happen in one pass, instead of building up an exclusion list.

eerhardt commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageDependencies.cs:
```diff
@@ -216,7 +216,7 @@ private void GetPackageAndFileDefinitions()
                     fileItem.SetMetadata(MetadataKeys.ResolvedPath, resolvedPath ?? string.Empty);
 
-                    if (NuGetUtils.IsApplicableAnalyzer(file, ProjectLanguage))
+                    if (NuGetUtils.IsApplicableAnalyzer(file, ProjectLanguage, excludedAnalyzers: null))
```
This change is no longer necessary.
